stix2/base.py

```diff
@@ -347,6 +360,61 @@ def _check_property(self, prop_name, prop, kwargs):
             for ref in kwargs[prop_name]:
                 self._check_ref(ref, prop, prop_name)
 
+    def _generate_id(self, kwargs):
+        required_prefix = self._type + "--"
+        namespace = uuid.UUID("00abedb4-aa42-466c-9c01-fed23315a9b7")
+
+        try:
+            properties_to_use = self._id_contributing_properties
+            if properties_to_use:
+                streamlined_object = {}
+                if "hashes" in kwargs and "hashes" in properties_to_use:
+                    possible_hash = self._choose_one_hash(kwargs["hashes"])
+                    if possible_hash:
+                        streamlined_object["hashes"] = possible_hash
+                for key in kwargs.keys():
+                    if key in properties_to_use and key != "hashes":
+                        if type(kwargs[key]) is dict:
+                            for otherKey in kwargs[key]:
+                                if isinstance(kwargs[key][otherKey], _STIXBase):
+                                    streamlined_object[key] = self._embed_obj_to_json(kwargs[key][otherKey])
+                                else:
+                                    streamlined_object[key] = kwargs[key]
+                        else:
+                            if isinstance(kwargs[key], _STIXBase):
+                                streamlined_object[key] = self._embed_obj_to_json(kwargs[key])
+                            else:
+                                streamlined_object[key] = kwargs[key]
+
+                if streamlined_object:
+                    data = canonicalize(str(streamlined_object), utf8=False)
+                    return required_prefix + str(uuid.uuid5(namespace, str(data)))
+            return None
+        except AttributeError:
```

Comment:
That's a big complicated chunk of code in the try-except.  I'd like to be able to look inside and understand what attribute you expect not to be there, but there's way too much junk in there to figure it out.  Can you narrow your try-except's to the minimal amount of coverage required?